File: test/services/jv-async/src/main/java/example/PetStoreAsyncImpl.java
package example;

import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

import petstore.api.Pet;
import petstore.api.PetStatus;

/**
 * Async server implementation attempting to implement the generated PetStore
 * interface with CompletableFuture<T> return types.
 *
 * This FAILS to compile because --jv-async-services=true only makes the
 * generated CLIENT return CompletableFuture<T>, but the generated server
 * interface (PetStore.java) still declares bare T return types. An async
 * server impl returning CompletableFuture<T> does not satisfy the generated
 * synchronous interface — reproducing D25.
 */
public final class PetStoreAsyncImpl implements petstore.api.PetStore {
    private final Map<Long, Pet> pets = new ConcurrentHashMap<>();
    private final AtomicLong nextId = new AtomicLong(1);

    // These methods return CompletableFuture<T>, but the generated PetStore
    // interface declares bare T return types. The @Override annotations will
    // cause javac to reject the file with:
    //   error: method does not override or implement a method from a supertype
    // and the return-type mismatch will surface as:
    //   error: addPet(petstore.api.petstore.addpet.In) in PetStoreAsyncImpl
    //   cannot override addPet(petstore.api.petstore.addpet.In) in PetStore
    //   return type CompletableFuture is not compatible with
    //   petstore.api.petstore.addpet.Out

    @Override
    public CompletableFuture<petstore.api.petstore.addpet.Out> addPet(
            petstore.api.petstore.addpet.In arg) {
        long id = nextId.getAndIncrement();
        Pet pet = new Pet(id, arg.name(), arg.status(), arg.tag());
        pets.put(id, pet);
        return CompletableFuture.completedFuture(
            new petstore.api.petstore.addpet.Out(pet));
    }

    @Override
    public CompletableFuture<petstore.api.petstore.getpet.Out> getPet(
            petstore.api.petstore.getpet.In arg) {
        Pet pet = pets.get(arg.id());
        if (pet == null) {
            throw new RuntimeException("Pet not found: " + arg.id());
        }
        return CompletableFuture.completedFuture(
            new petstore.api.petstore.getpet.Out(pet));
    }

    @Override
    public CompletableFuture<petstore.api.petstore.listpets.Out> listPets(
            petstore.api.petstore.listpets.In arg) {
        List<Pet> sorted = pets.values().stream()
            .sorted(Comparator.comparingLong(Pet::id))
            .toList();
        return CompletableFuture.completedFuture(
            new petstore.api.petstore.listpets.Out(sorted));
    }

    @Override
    public CompletableFuture<petstore.api.petstore.deletepet.Out> deletePet(
            petstore.api.petstore.deletepet.In arg) {
        Pet removed = pets.remove(arg.id());
        return CompletableFuture.completedFuture(
            new petstore.api.petstore.deletepet.Out(removed != null));
    }
}

File: test/services/kt-errors-async/src/main/kotlin/example/PetStoreAsyncImpl.kt
package example

import baboon.runtime.shared.Either

/**
 * Errors-mode async server implementation for the petstore-errors model.
 *
 * This attempts to implement the generated PetStore interface using:
 *   - `suspend fun` methods (async mode: --kt-async-services=true)
 *   - Either<Err, Out> return types (errors mode: --service-result-no-errors=false
 *     --service-result-type=Either --service-result-pattern="<$error,$success>")
 *
 * This FAILS to compile because KtServiceWiringTranslator emits the server
 * dispatch wiring using a plain (non-suspend) lambda for IBaboonServiceRt.flatMap:
 *
 *   rt.flatMap<...>(input) { v ->
 *     impl.addPet(v)      // ERROR: suspend call inside non-suspend lambda
 *   }
 *
 * When --kt-async-services=true, the generated PetStore interface declares
 * `suspend fun addPet(...)`, so calling it inside the non-suspend rt.flatMap
 * lambda body causes a kotlinc error:
 *   "Suspend function 'suspend fun addPet(...)' should be called only from
 *    a coroutine or another suspend function."
 *
 * This reproduces defect D26 (errors+async wiring): the same root cause as D25
 * (no-errors+async), but in the errors code path of KtServiceWiringTranslator
 * (generateErrorsJsonMethod / generateErrorsUebaMethod).
 *
 * Expected kotlinc error:
 *   Suspend function 'suspend fun addPet(arg: In): Either<Err, Out>'
 *   should be called only from a coroutine or another suspend function.
 */
class PetStoreAsyncImpl : petstore.api.PetStore {
    private val pets = mutableMapOf<Long, petstore.api.Pet>()
    private var nextId = 1L

    fun reset() {
        pets.clear()
        nextId = 1L
    }

    // These methods are both `suspend` (async mode) AND return Either<Err, Out>
    // (errors mode). The generated PetStore interface declares:
    //   suspend fun addPet(arg: petstore.api.petstore.addpet.In):
    //       Either<petstore.api.petstore.addpet.Err, petstore.api.petstore.addpet.Out>
    // The generated PetStoreWiring will then try to call impl.addPet(v) inside
    // a non-suspend rt.flatMap lambda, causing a kotlinc compile error (D26).

    override suspend fun addPet(arg: petstore.api.petstore.addpet.In): Either<petstore.api.petstore.addpet.Err, petstore.api.petstore.addpet.Out> {
        val id = nextId++
        val pet = petstore.api.Pet(
            id = id,
            name = arg.name,
            status = arg.status,
            tag = arg.tag
        )
        pets[id] = pet
        return Either.Right(petstore.api.petstore.addpet.Out(pet = pet))
    }

    override suspend fun getPet(arg: petstore.api.petstore.getpet.In): Either<petstore.api.petstore.getpet.Err, petstore.api.petstore.getpet.Out> {
        val pet = pets[arg.id]
        return if (pet != null) {
            Either.Right(petstore.api.petstore.getpet.Out(pet = pet))
        } else {
            Either.Left(petstore.api.petstore.getpet.Err(
                problem = petstore.api.PetError(code = 404, message = "Pet not found: id=${arg.id}")
            ))
        }
    }
}
